drivers: net: lan8651: Provide fixes for OA TC6 LAN865x driver

lmajewski opened the pull request:
This patch set provides fixes for LAN865x OA TC6 driver as a follow up of https://github.com/zephyrproject-rtos/zephyr/pull/63614

Most notable fixes:
- Improve stability with refactoring and proper configuration of semaphores
- Avoid sending extra chunk when packet size equals to multiple size of chunks
- Add and use RMW operation
- Add saturation code for RCA and TXC when read from OA_BUFSTS register
- Add return code check in interrupt thread to avoid using stalled RCA and TXC data
- Do not send chunk when no data provided.
- Use ZARFE bit from OA_CONFIG to avoid fragments with packet header

Commit:
drivers: net: tc6: Add read,modify and write operation on control regs

The OA TC6 driver requires some bits manipulations in control registers.
Up till now - it has been implemented as an explicit set of read and write
registers' operations.
One good example would be the oa_tc6_set_protected_ctrl() implementation,
which used such scheme.

This patch brings dedicated function for this operation; oa_tc6_reg_rmw().
The aforementioned oa_tc6_set_protected_ctrl() function now uses it
internally.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: Force LAN865x to start received frames at first chunk word

Zephyr's network stack has issues with network IP header split across
fragments.
To alleviate this problem, the frame would be now aligned to first byte
of the chunk. This would ensure that the header is stored at one network
buffer fragment (as we explicitly set its size to 64B).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: tc6: Saturate RCA and RBA when read directly from OA_BUFSTS

The RCA and RBA fields in OA_BUFSTS register are stored with 8 bits each.

On the other hand, when one receives those values in footer, the value
is saturated to 5 bits due to 32 bit size constrain of the footer itself.

To avoid any mismatches - the values read from OA_BUFSTS are saturated
as well.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: lan865x: Do not read chunks if protected read error detected

The oa_tc6_update_buf_info() function returns error code when read
(protected or not) of OA_BUFSTS has been detected.

In that situation - one shall re-start the interrupt thread handling
(and hence correctly re-read value of this register) than use old
(stalled) RCA(RBA) data to read chunks (which may result in lockup).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: tc6: Do not send extra last chunk when not required

Current code sends extra, last chunk, when packet's size is a multiple
of chunk size (i.e. 64 bytes).

This patch fixes this issue by checking this corner case - i.e. if
the modulo division equals to zero.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: tc6: Avoid sending packet when no data provided

This commit adds check if the packet to be sent has any data.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: lan865x: Fix use of RX/TX semaphore in read chunks function

This commit ensures that whole receive function (called from interrupt
handler) is protected by the RX/TX semaphore.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
drivers: net: lan865x: Fix initialization of semaphores

This change fixes semaphores' definition. To be more specific - the
limit was wrongly set to UINT_MAX. With those changes - the 'tx_rx_sem'
now assures that only one execution path (i.e. receiving or sending
data) is executed at a time.

Moreover, the change in 'int_sem' now assures that this semaphore limit,
when reached, is saturated.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jukkar commented on drivers/ethernet/eth_lan865x.c:
```diff
@@ -373,16 +373,32 @@ static void lan865x_read_chunks(const struct device *dev)
+		 * Allow only TX, so net stack have time to process
+		 * already received data.
+		 */
+		k_msleep(cfg->timeout / 10);
```
I am not a big fan of adding sleeps in the code, sometimes they work but it is difficult to decide how long to sleep etc. 
Would it be possible to change this so that the TX side informs when the RX side can proceed, perhaps via a mutex with a timeout?

lmajewski replied:
@jukkar - the time here is related to the timeout used previously with RX buffer allocation (`cfg->timeout`). The assumption here is that after 1/10 of time which we waited for RX packet to be allocated there shall be some memory available (as some packets are "consumed" by network stack).

Maybe, I shall change the commit message of this particular commit (as it seems to be misleading)? The problem which I've been solving is when I do try to send too much data in packets from HOST to LAN8651. As we do have some fixed number of available memory for RX chunks (rather smaller number) the "stall" caused by no memory in RX pool happens from time to time. I could "rely" on LAN8651 HW feature as it overwrites the buffer, but presented here solution seems to be simpler and less intrusive.

lmajewski replied:
I've discarded this commit as it brought more issues (with the sleep) than benefits. The driver shall behave in a robust way without it.

jukkar reviewed (commented):
Looks ok. Please adjust the commit subject and add `drivers: ...` prefix as all of these are driver changes.

Commit:
drivers: net: tc6: Provide separate function to check IC status

The code to handle OA TC6 compliant device's status is generic and can
be moved to device agnostic driver (oa_tc6.c).

Moreover, the original code has been augmented with LOG_WRN() messages.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jukkar commented:
Please adjust the **commit** subjects (not the PR) and then force push the commits.

lmajewski commented:
> Please adjust the **commit** subjects (not the PR) and then force push the commits.

I've adjusted the commit messages as requested :-)

jukkar reviewed (approved):
LGTM
